File: java/src/main/java/com/mlt/metadata/stream/StreamMetadataDecoder.java
package com.mlt.metadata.stream;

import me.lemire.integercompression.IntWrapper;

public class StreamMetadataDecoder {

  public static StreamMetadata decode(byte[] tile, IntWrapper offset) {
    var streamMetadata = StreamMetadata.decode(tile, offset);
    /* Currently morton can't be combined with RLE only with delta */
    if (streamMetadata.logicalLevelTechnique1().equals(LogicalLevelTechnique.MORTON)) {
      return MortonEncodedStreamMetadata.decodePartial(streamMetadata, tile, offset);
    }
    /* Boolean rle doesn't need additional information */
    else if ((LogicalLevelTechnique.RLE.equals(streamMetadata.logicalLevelTechnique1())
            || LogicalLevelTechnique.RLE.equals(streamMetadata.logicalLevelTechnique2()))
        && !PhysicalLevelTechnique.NONE.equals(streamMetadata.physicalLevelTechnique())) {
      return RleEncodedStreamMetadata.decodePartial(streamMetadata, tile, offset);
    }

    return streamMetadata;
  }
}

File: js/src/metadata/stream/StreamMetadataDecoder.ts
import { IntWrapper } from '../../decoder/IntWrapper';
import { StreamMetadata } from './StreamMetadata';
import { LogicalLevelTechnique } from './LogicalLevelTechnique';
import { PhysicalLevelTechnique } from './PhysicalLevelTechnique';
import { MortonEncodedStreamMetadata } from './MortonEncodedStreamMetadata';
import { RleEncodedStreamMetadata } from './RleEncodedStreamMetadata';

export class StreamMetadataDecoder {
    public static decode(tile: Uint8Array, offset: IntWrapper): StreamMetadata {
        const streamMetadata = StreamMetadata.decode(tile, offset);
        if (streamMetadata.logicalLevelTechnique1() === LogicalLevelTechnique.MORTON) {
            return MortonEncodedStreamMetadata.decodePartial(streamMetadata, tile, offset);
        } else if (
            (LogicalLevelTechnique.RLE === streamMetadata.logicalLevelTechnique1() ||
                LogicalLevelTechnique.RLE === streamMetadata.logicalLevelTechnique2()) &&
            PhysicalLevelTechnique.NONE !== streamMetadata.physicalLevelTechnique()
        ) {
            return RleEncodedStreamMetadata.decodePartial(streamMetadata, tile, offset);
        }

        return streamMetadata;
    }
}
